src/gtclang/Frontend/GlobalVariableParser.cpp

```diff
@@ -149,16 +149,29 @@ void GlobalVariableParser::parseGlobals(clang::CXXRecordDecl* recordDecl) {
             << init->getType().getAsString() << name;
       };
 
-      if(IntegerLiteral* il = dyn_cast<IntegerLiteral>(init)) {
+      // demotion to integer (`double ->12<-' would dyncast to int)
+      if(dyn_cast<IntegerLiteral>(init) != nullptr && typeKind == dawn::sir::Value::Integer) {
+        IntegerLiteral* il = dyn_cast<IntegerLiteral>(init);
         std::string valueStr = il->getValue().toString(10, true);
         value->setValue((int)std::atoi(valueStr.c_str()));
         DAWN_LOG(INFO) << "Setting default value of '" << name << "' to '" << valueStr << "'";
 
-      } else if(FloatingLiteral* fl = dyn_cast<FloatingLiteral>(init)) {
-        llvm::SmallVector<char, 10> valueVec;
-        fl->getValue().toString(valueVec);
-        std::string valueStr(valueVec.data(), valueVec.size());
-        value->setValue((double)std::atof(valueStr.c_str()));
+      // this slightly unelegant procedure is needed since FloatingLiteral does not cast from
+      // expressions without trailing do (e.g. `12.' would cast, `12' wouldn't.)
+      } else if((dyn_cast<FloatingLiteral>(init) != nullptr || dyn_cast<IntegerLiteral>(init) != nullptr)
+          && typeKind == dawn::sir::Value::Double) {
+        IntegerLiteral* il = dyn_cast<IntegerLiteral>(init);
+        FloatingLiteral* fl = dyn_cast<FloatingLiteral>(init);
+        std::string valueStr;
+        if (fl != nullptr) {
+          llvm::SmallVector<char, 10> valueVec;
```

Comment:
just a small improvement suggestion: putting magic number 10 into a const int member of GlobalVariableParser and maybe making it a 20 (3.1415926535 breaks it already).

Reply:
update: for now we keep it, but plan to refactor magic numbers away from the whole code sometime